### PR TITLE
Spherical padding and faster tests

### LINKED ISSUE #8: Translate between (-180 -> 180) and (0 -> 360) longitudes
CDO automagically deals with the translation between (-180 -> 180) and (0 -> 360) longitude coordinates.
This is a relatively minor feature to implement, and can make it a lot nicer for users.

### LINKED ISSUE #9: Improve results on data boundaries
Currently the regridding methods are oblivious to the fact that the word wraps around. This can lead to errors on the edges of the input data (e.g. latitude 0 / latitude -180).

### LINKED ISSUE #42: Improve performance of conservative routine
> We can lessen the performance penalty significantly by doing something like `notnull.any([non_grid_dims])` here, in which case we track the nan fraction as any batch slices that have valid data. Could be a reasonable tradeoff, or a configurable argument.

_Originally posted by @slevang in https://github.com/EXCITED-CO2/xarray-regrid/pull/39#discussion_r1677154166_

### PR DESCRIPTION
**Pull request by slevang**
- [x] closes #8 
- [x] closes #9 

## Spherical padding

This makes an initial attempt at building in some automatic logic for better handling the boundaries of a spherical domain with appropriate padding, as well as shifting the longitude source grid to match the target as needed.

## Faster tests

Made some modifications and general cleanup of the existing tests. Mainly making the datasets dask-backed so the regridding functions are lazy for all the attribute checks. Also got rid of some unneeded copies, compute calls, etc. Tests run in about 17 seconds now.

With the addition of the spherical padding logic I was also able to remove some aspects of the tests where we avoided checking for good matches at the boundaries.

## Conservative NaN tracking

I did some benchmarking and the version of NaN tracking where we track independently across non-grid dimensions as well doesn't actually have much of a run-time penalty. Therefore I went ahead and made this small change. The only issue is that it can induce a very large memory penalty if the input data isn't chunked, because the weights arrays can become huge if they are broadcast with all the other input dimensions. I added a note to the docstring about this but perhaps we want a more formal warning, or to explicitly add some chunking in certain cases?

## TODO / open questions

- [x] tests added
- [x] fix typing
- [x] add examples to docstrings
- [ ] add args to make the padding behavior configurable?

### CODE REVIEW

File: src/xarray_regrid/methods/conservative.py [hunk 1]
```diff
@@ -161,16 +159,13 @@ def apply_weights(
...
     skipna: bool,
-    non_grid_dims: list[Hashable],
 ) -> tuple[xr.DataArray, xr.DataArray]:
     """Apply the weights to convert data to the new coordinates."""
     coord_map = {f"target_{coord}": coord}
     weights_norm = weights.copy()
 
     if skipna:
         notnull = da.notnull()
-        if non_grid_dims:
```
Comment (slevang) [inline src/xarray_regrid/methods/conservative.py:172]: This was the somewhat arbitrary reduction step I had added to the NaN tracking logic before, to consider grid points where any non-grid sample was null. Taking this out allows us to deal with isolated NaNs in a particular timestamp for example.

File: src/xarray_regrid/utils.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def create_lat_lon_coords(grid: Grid) -> tuple[np.ndarray, np.ndarray]:
             grid.south, grid.north + grid.resolution_lat, grid.resolution_lat
         )
 
-    if np.remainder((grid.north - grid.south), grid.resolution_lat) > 0:
+    if np.remainder((grid.east - grid.west), grid.resolution_lat) > 0:
```
Comment (slevang) [inline src/xarray_regrid/utils.py:78]: Unrelated but I think this was not intended?

File: src/xarray_regrid/utils.py [hunk 3]
```diff
@@ -235,3 +235,127 @@ def call_on_dataset(
...
+    add a single wraparound padding column if the domain is global and the east
+    or west edges of the target lie outside the source grid centers.
+
+    For latitude, add a single value at each pole computed as the mean of the last
+    row for global source grids where the first or last point lie equatorward of 90.
+    """
+
+    orig_chunksizes = obj.chunksizes
+
+    # If the source coord fully covers the target, don't modify them
```
Comment (slevang) [inline src/xarray_regrid/utils.py:281]: This got pretty complex, happy to take any suggestions for improvement, and we could probably factor things out into more testable components.

File: tests/test_regrid.py [hunk 4]
```diff
@@ -97,46 +72,60 @@ def conservative_sample_grid():
...
-def test_conservative_regridder(conservative_input_data, conservative_sample_grid):
+@pytest.mark.parametrize("method", ["linear", "nearest"])
+def test_basic_regridders_ds(
+    sample_input_data, sample_grid_ds, cdo_comparison_data, method
+):
+    """Test the dataset regridders (except conservative)."""
+    regridder = getattr(sample_input_data.regrid, method)
+    ds_regrid = regridder(sample_grid_ds)
+    ds_cdo = cdo_comparison_data[method]
+    xr.testing.assert_allclose(ds_regrid, ds_cdo, rtol=0.002, atol=2e-5)
```
Comment (slevang) [inline tests/test_regrid.py:83]: After padding I was able to remove the edge selection in these and still get the tests to pass with a slight reduction in the tolerance.

File: tests/test_regrid.py [hunk 5]
```diff
@@ -235,16 +214,14 @@ def test_conservative_nan_thresholds_against_xesmf():
...
-        data_regrid = ds.copy().regrid.conservative(
+        data_regrid = ds.regrid.conservative(
             target_dataset, skipna=True, nan_threshold=nan_threshold
         )
-        data_esmf = regridder(
-            ds.copy(), keep_attrs=True, na_thres=nan_threshold, skipna=True
-        )
-        assert (data_regrid.isnull() == data_esmf.isnull()).mean().values > 0.995
+        data_esmf = regridder(ds, keep_attrs=True, na_thres=nan_threshold, skipna=True)
+        xr.testing.assert_equal(data_regrid.isnull(), data_esmf.isnull())
```
Comment (slevang) [inline tests/test_regrid.py:224]: NaN thresholding now matches xesmf exactly with the new padding.

Comment (BSchilperoort) [reply tests/test_regrid.py:222]: Nice!

File: tests/test_regrid.py [hunk 6]
```diff
@@ -27,20 +27,22 @@
 @pytest.fixture(scope="session")
 def sample_input_data() -> xr.Dataset:
     ds = xr.open_dataset(DATA_PATH / "era5_2m_dewpoint_temperature_2000_monthly.nc")
-    return ds.isel(time=slice(0, N_TIMESTAMPS)).persist()
+    # slice after compute due to current xarray bug: https://github.com/pydata/xarray/issues/8909
+    # then convert to dask so regridding is lazy on attribute tests
+    return ds.compute().isel(time=slice(0, N_TIMESTAMPS)).chunk()
```
Comment (BSchilperoort) [inline tests/test_regrid.py]: From reading that issue it seems like we can also avoid the bug by installing netCDF4 as dependency? We can add it to the dev dependencies.

File: src/xarray_regrid/utils.py (hunk 2 above)
Comment (BSchilperoort) [reply src/xarray_regrid/utils.py:78]: Yeah looks like a bug. Nice catch

File: src/xarray_regrid/methods/conservative.py [hunk 7]
```diff
@@ -66,7 +66,7 @@ def conservative_regrid(
     # Attempt to infer the latitude coordinate
     if latitude_coord is None:
         for coord in data.coords:
-            if str(coord).lower() in ["lat", "latitude"]:
+            if str(coord).lower().startswith("lat"):
```
Comment (BSchilperoort) [inline src/xarray_regrid/methods/conservative.py:69]: I had purposefully made it check only if the coordinate is `lat` or `latitude`. If it's something different, the user can specify it.

Comment (slevang) [reply src/xarray_regrid/methods/conservative.py:69]: Sure, happy to revert this. I have occasionally seen for example `long` as a coordinate, but this could theoretically catch unwanted similar names as well.

File: src/xarray_regrid/utils.py (hunk 3 above)
Comment (BSchilperoort) [reply src/xarray_regrid/utils.py:281]: Yeah I am having some trouble going through the logic. It might be nice to separate the routines for latitude and longitude in their own functions, and add a short example in the docstrings of the functions on how the coordinate is modified.

File: tests/test_regrid.py (hunk 4 above)
Comment (BSchilperoort) [reply tests/test_regrid.py:81]: Awesome! :smile:

File: src/xarray_regrid/utils.py (hunk 3 above)
Comment (slevang) [reply src/xarray_regrid/utils.py]: Did some refactoring and added descriptive examples to the docstrings, hopefully a little better now.

File: src/xarray_regrid/utils.py [hunk 8]
```diff
@@ -235,3 +240,151 @@ def call_on_dataset(
...
+        obj = xr.concat([obj, north_pole], dim=lat_coord)  # type: ignore
+        obj.coords[lat_coord].values[-1] = polar_lat
+
+    return obj
+
+
+def format_lon(
+    obj: xr.DataArray | xr.Dataset, target: xr.Dataset, formatted_coords: dict[str, str]
+) -> xr.DataArray | xr.Dataset:
+    """Format the longitude coordinate by shifting the source grid to line up with
```
Comment (BSchilperoort) [inline src/xarray_regrid/utils.py:331]: Nice!

File: src/xarray_regrid/utils.py [hunk 9]
```diff
@@ -235,3 +240,151 @@ def call_on_dataset(
...
+
+    return obj
+
+
+def format_lat(
+    obj: xr.DataArray | xr.Dataset,
+    target: xr.Dataset,  # noqa ARG001
+    formatted_coords: dict[str, str],
+) -> xr.DataArray | xr.Dataset:
+    """If the latitude coordinate is inferred to be global, defined as having
```
Comment (BSchilperoort) [inline src/xarray_regrid/utils.py:284]: Efficient reuse of ESMF's logic, no need to reinvent the wheel :smile:

File: tests/test_format.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    # Should add wraparound and polar padding rows/columns
+    assert formatted.latitude[0] == -90
+    assert formatted.latitude[-1] == 90
+    assert formatted.longitude[0] == -1
+    assert formatted.longitude[-1] == 361
+    assert (formatted.longitude.diff("longitude") == 2).all()
+
+
+def test_360_to_180():
```
Comment (BSchilperoort) [inline tests/test_format.py:65]: Would it be useful here to parameterize the test by `dx_source` (and maybe dx_target)?

Also, these tests only test for regularly spaced grids, what if a grid has irregular spacing?

File: tests/test_regrid.py [hunk 11]
```diff
@@ -212,19 +199,9 @@ def test_conservative_nan_thresholds_against_coarsen(nan_threshold):
...
-    return True
-
-
-@pytest.mark.skipif(not xesmf_available(), reason="xesmf required")
+@pytest.mark.skipif(xesmf is None, reason="xesmf required")
 def test_conservative_nan_thresholds_against_xesmf():
-    import xesmf as xe
-
-    ds = xr.tutorial.open_dataset("ersstv5").sst.compute()
+    ds = xr.tutorial.open_dataset("ersstv5").sst.compute().isel(time=[0])
```
Comment (BSchilperoort) [inline tests/test_regrid.py:204]:
```suggestion
    ds = xr.tutorial.open_dataset("ersstv5")["sst"].isel(time=[0]).compute()
```
First selecting and then computing should be a little more efficient right? But looking at the dataset's size it probably won't matter much.

File: src/xarray_regrid/utils.py (hunk 3 above)
Comment (BSchilperoort) [reply src/xarray_regrid/utils.py]: It's more clear to me now! Thanks

**Review by BSchilperoort — APPROVED** [review]
Thanks for making these improvements!

It would probably be good to add a single workflow runner [with micromamba](https://github.com/mamba-org/setup-micromamba), so that the comparison with ESMF will run on the CI. I could do this in a separate PR.

> The only issue is that it can induce a very large memory penalty if the input data isn't chunked, because the weights arrays can become huge if they are broadcast with all the other input dimensions. I added a note to the docstring about this but perhaps we want a more formal warning, or to explicitly add some chunking in certain cases?

This probably is quite unintuitive to users, was the reason for removing this tracking? mostly just simplifying the code? If it does allow for a much lower memory footprint (without users needing to improve their chunking) it could be interesting to leave the non-grid-dims tracking in.

> Made some modifications and general cleanup of the existing tests. Mainly making the datasets dask-backed so the regridding functions are lazy for all the attribute checks. Also got rid of some unneeded copies, compute calls, etc. Tests run in about 17 seconds now.

Awesome!

File: tests/test_regrid.py (hunk 11 above)
Comment (slevang) [reply tests/test_regrid.py:204]: This was the weird bug with current version of `netcdf4`. It adds very little time but yeah hopefully they fix it or we could switch to `h5netcdf` in the env. I'll leave it as is for now.

File: tests/test_format.py (hunk 10 above)
Comment (slevang) [reply tests/test_format.py:65]: These would be good additions, I can follow up with another PR to cover more test cases.

**Comment by slevang** [comment]
> This probably is quite unintuitive to users, was the reason for removing this tracking? mostly just simplifying the code? If it does allow for a much lower memory footprint (without users needing to improve their chunking) it could be interesting to leave the non-grid-dims tracking in.

It does simplify the code. But primarily, getting rid of this aggregation of the valid frac over non grid dims allows us to truly track isolated NaN fractions, so that you can have NaN in just a single time slice for example and get the correct output. I think what I have now is the most sensible baseline option, but I have some other ideas for improving conservative performance so I'll follow up in a different branch to avoid adding too much to this PR.

**Comment by slevang** [comment]
I made a few small changes/refactors which I guess cancelled out your previous review.

File: src/xarray_regrid/utils.py [hunk 12]
```diff
@@ -388,3 +369,50 @@ def coord_is_covered(
...
+
+@overload
+def ensure_monotonic(obj: xr.DataArray, coord: Hashable) -> xr.DataArray: ...
+
+
+@overload
+def ensure_monotonic(obj: xr.Dataset, coord: Hashable) -> xr.Dataset: ...
+
+
+def ensure_monotonic(
```
Comment (BSchilperoort) [inline src/xarray_regrid/utils.py:382]: I was also working on this exact function for #46, see https://github.com/xarray-contrib/xarray-regrid/pull/46#discussion_r1751955505.

You can use the following to make the code more simple:
```python
obj.indexes[coord].is_monotonic_increasing
obj.indexes[coord].is_unique
```

I'm also not sure what is more efficient: sorting/dropping duplicates one coord at a time, or tackling all in one go. My approach was this;
```py
def ensure_monotonic(
    obj: xr.Dataset | xr.DataArray, coords: list[Hashable]
) -> xr.Dataset | xr.DataArray:
    """Sort coordinates that do not monotonically increase.

    If no sorting needs to take place, the data is not sorted."""
    unsorted_coords = [
        coord for coord in coords if not obj.indexes[coord].is_monotonic_increasing
    ]
    if len(unsorted_coords) > 0:
        obj = obj.sortby(unsorted_coords)

    duplicate_coords = [coord for coord in coords if not obj.indexes[coord].is_unique]
    if len(duplicate_coords) > 0:
        obj = obj.drop_duplicates(duplicate_coords)

    return obj
```

File: src/xarray_regrid/methods/conservative.py [hunk 13]
```diff
@@ -72,8 +72,10 @@ def conservative_regrid(
 
     # Make sure the regridding coordinates are sorted
     coord_names = [coord for coord in target_ds.coords if coord in data.coords]
-    target_ds_sorted = target_ds.sortby(coord_names)
-    data = data.sortby(list(coord_names))
+    target_ds_sorted = target_ds.copy()
```
Comment (BSchilperoort) [inline src/xarray_regrid/methods/conservative.py:75]:
```suggestion
    target_ds_sorted = xr.Dataset(target_ds.coords)
```
We don't need the values of the target dataset, only the coordinates. This could reduce any potential copies of the target_ds data.

**Review by BSchilperoort — APPROVED** [review]
Thanks for the improvements! :smile:
